### PR TITLE
Added device API support

### LINKED ISSUE #3: Add /device API support
Requires https://github.com/punica/punica/issues/7

### PR DESCRIPTION
**Pull request by mrNedas**
Closes #3

### CODE REVIEW

File: punica.js [hunk 1]
```diff
@@ -67,6 +67,30 @@ class Device extends EventEmitter {
     });
   }
 
+  /**
```
Comment (GiedriusM) [inline punica.js]: Šiuo metu yra toks niuansas, kad /endpoints ir /devices naudoja skirtingus ID, todėl šita funkciją turėtų neveikti. Geriau dabar išimti.

File: punica.js [hunk 2]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * Sends request to get all registered devices entries.
+   * @returns {Promise} Promise with a list of endpoints
```
Comment (GiedriusM) [inline punica.js]: * Promise duplikojasi
* Ne endpoints, o device registracijos

Siūlyčiau kažką tokio: `@returns {Promise} which fulfills with a list of registered devices`

File: punica.js [hunk 3]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * @example
+   * service.getDevicesEntries().then((resp) => {
+   *   // resp = [{"psk_id": "cHNraWQx", "uuid": "ABC"}, ...]
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   */
+  getDevicesEntries() {
```
Comment (GiedriusM) [inline punica.js]: Kad nesimaišytų sų /endpoints device informacija, siūliau pervadinti į `getRegisteredDevices`, o `getDevices` aprašyme įdėti žodelį connected.

File: punica.js [hunk 4]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+          reject(dataAndResponse.resp.statusCode);
+        }
+      }).catch((err) => {
+        reject(err);
+      });
+    });
+  }
+
+  /**
+   * Sends request to get device's PSK ID.
```
Comment (GiedriusM) [inline punica.js]: get device registration info/entry

File: punica.js [hunk 5]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * @returns {Promise} Promise with device's PSK ID
+   * @example
+   * const name = 'DEF';
+   * service.getRegisteredDevicePskId(name).then((resp) => {
+   *   // resp = {"psk_id": "cHNraWQy", "uuid": "DEF"}
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   * @param {string} name - Device name
```
Comment (GiedriusM) [inline punica.js]: Pakeitėm kad UUID, o ne name.

File: punica.js [hunk 6]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   */
+  getDeviceEntry(name) {
```
Comment (GiedriusM) [inline punica.js]: getRegisteredDevice?

File: punica.js [hunk 7]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * @example
+   * const entry = {"psk":"cHNrMQ==","psk_id":"cHNraWQx"};
+   * service.registerDeviceEntry(entry).then((resp) => {
+   *   // resp = '' (Successfully registered)
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   * @param {object} entry - A JSON object representing the device entry
+   */
+  registerDeviceEntry(entry) {
```
Comment (GiedriusM) [inline punica.js]: Siūlau pervadinti į createRegisteredDevice.

File: punica.js [hunk 8]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+          reject(dataAndResponse.resp.statusCode);
+        }
+      }).catch((err) => {
+        reject(err);
+      });
+    });
+  }
+
+  /**
+   * Sends request to register devices.
```
Comment (GiedriusM) [inline punica.js]: ... to register a new device.

File: punica.js [hunk 9]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * @returns {Promise} Promise which fulfills when devices are registered successfully
+   * @example
+   * const entry = {"psk":"cHNrMQ==","psk_id":"cHNraWQx"};
+   * service.registerDeviceEntry(entry).then((resp) => {
+   *   // resp = '' (Successfully registered)
```
Comment (GiedriusM) [inline punica.js]: Serveris turėtų grąžinti visą info, t.y. psk, psk_id, sugeneruotą uuid.

File: punica.js [hunk 10]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * const entry = {"psk":"cHNrMQ==","psk_id":"cHNraWQa"};
+   * service.updateDeviceEntry(name, entry).then((resp) => {
+   *   // resp = '' (Successfully edited)
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   * @param {string} name - Device name
+   * @param {object} entry - A JSON object with updated 'psk' and 'psk_id'
+   */
+  updateDeviceEntry(name, entry) {
```
Comment (GiedriusM) [inline punica.js]: Siūlau pervadint updateRegistaredDevice. Taip pat ne name, o uuid.

File: punica.js [hunk 11]
```diff
@@ -637,6 +661,136 @@ class Service extends EventEmitter {
...
+   * @example
+   * const name = 'DEF';
+   * service.removeDeviceEntry(name).then((resp) => {
+   *   // resp = '' (Successfully removed)
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   * @param {string} name - Device name
+   */
+  removeDeviceEntry(name) {
```
Comment (GiedriusM) [inline punica.js]: Siūlau pervadinti į `removeRegisteredDevice`.

File: README.md [hunk 12]
```diff
@@ -143,6 +144,21 @@ const restAPI = require('punica');
 const service = new restAPI.Service(serviceOptions);
 const device = new restAPI.Device(service, 'deviceId');
 ```
+<a name="Device+getEntry"></a>
+
+### device.getEntry() ⇒ <code>Promise</code>
+Sends request to get device's entry.
+
+**Kind**: instance method of [<code>Device</code>](#Device)  
+**Returns**: <code>Promise</code> - Promise with device's PSK ID
```
Comment (taubel) [inline README.md]: Doesn't return only device's PSK ID. Should be renamed to device entry.

File: README.md [hunk 13]
```diff
@@ -478,6 +499,107 @@ service.getDevices().then((resp) => {
   // err - exception message object or status code
 });
 ```
+<a name="Service+getDevicesEntries"></a>
+
+### service.getDevicesEntries() ⇒ <code>Promise</code>
+Sends request to get all registered devices entries.
+
+**Kind**: instance method of [<code>Service</code>](#Service)  
+**Returns**: <code>Promise</code> - Promise with a list of endpoints
```
Comment (taubel) [inline README.md]: I think `Promise with a list of device entries` would be more uniform throughout the readme

File: README.md [hunk 14]
```diff
@@ -478,6 +499,107 @@ service.getDevices().then((resp) => {
...
+
+### service.getDevicesEntries() ⇒ <code>Promise</code>
+Sends request to get all registered devices entries.
+
+**Kind**: instance method of [<code>Service</code>](#Service)  
+**Returns**: <code>Promise</code> - Promise with a list of endpoints  
+**Example**  
+```js
+service.getDevicesEntries().then((resp) => {
+  // resp = [{"psk_id": "cHNraWQx", "uuid": "ABC"}, ...]
```
Comment (taubel) [inline README.md:497]: Device UUID is now generated by the server, so all examples of UUID's in requests, responses etc. should follow the pattern `123e4567-e89b-12d3-a456-426655440000` or simmilar

File: README.md [hunk 15]
```diff
@@ -478,6 +499,107 @@ service.getDevices().then((resp) => {
...
+**Returns**: <code>Promise</code> - Promise with device's PSK ID  
+
+| Param | Type | Description |
+| --- | --- | --- |
+| name | <code>string</code> | Device name |
+
+**Example**  
+```js
+const name = 'DEF';
+service.getRegisteredDevicePskId(name).then((resp) => {
```
Comment (taubel) [inline README.md]: I would suggest `getRegisteredDeviceEntry`

File: tests/punica.test.js [hunk 16]
```diff
@@ -599,6 +629,155 @@ describe('Rest API interface', () => {
...
+          .catch((err) => {
+            expect(typeof err).to.equal('object');
+            done();
+          });
+      });
+    });
+
+    describe('registerDeviceEntry function', () => {
+      const data = { psk: 'cHNrMQ==', psk_id: 'cHNraWQx' };
+      it('should fulfill promise if registration is successfull (201)', () => {
```
Comment (taubel) [inline tests/punica.test.js]: There have been some changes to the API lately. Most importantly, UUID is generated by the server and is returned in the response of the POST request. The current punica API documentation does not reflect this, and will be fixed soon. You later can use the returned UUID during tests where you edit/delete/get a specific device entry.

**Review by taubel — CHANGES_REQUESTED** [review]

File: punica.js [hunk 17]
```diff
@@ -637,6 +637,136 @@ class Service extends EventEmitter {
...
+   * const entry = {"psk":"cHNrMQ==","psk_id":"cHNraWQa"};
+   * service.updateDeviceEntry(uuid, entry).then((resp) => {
+   *   // resp = '' (Successfully edited)
+   * }).catch((err) => {
+   *   // err - exception message object or status code
+   * });
+   * @param {string} uuid - Device uuid
+   * @param {object} entry - A JSON object with updated 'psk' and 'psk_id'
+   */
+  updateRegistaredDevice(uuid, entry) {
```
Comment (GiedriusM) [inline punica.js]: Typo: should be Registered instead of Registared.

**Review by GiedriusM — CHANGES_REQUESTED** [review]

**Review by GiedriusM — APPROVED** [review]

**Pull request merged**